ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,297 @@
+            return await self._get_variable_from_session(resource)
+
+        root_resource = await self._get_root_resource(resource)
+        if resource.get_id() == root_resource.get_id():
+            await self._add_variable_to_session(resource, "root_resource")
+            return "root_resource"
+
+        parent = await resource.get_parent()
+        if not await self._var_exists(parent):
+            await self.add_variable(parent)
+
+        selector = await self._get_selector(resource)
+        name = await self._generate_name(resource)
+        await self._add_action_to_session(
+            resource,
+            rf"""
+        {name} = {selector}""",
+            ActionType.UNDEF,
+        )
+        await self._add_variable_to_session(resource, name)
+        return name
+
+    async def delete_action(self, resource: Resource, action: str) -> None:
+        """
+        :param action:
+        """
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        # TODO: do we really need to delete an action from the script?
+        for idx, script_action in enumerate(session.actions):
+            if script_action.action == action:
+                session.actions.pop(idx)
+
+    async def get_script(self, resource: Resource) -> str:
+        """
+        :return script:
+        """
+        root_resource = await self._get_root_resource(resource)
+        return self._get_script(root_resource.get_id())
+
+    async def get_all_of_type(self, resource: Resource, target_type: ActionType) -> str:
+        """
+        :param target_type:
+        :return script:
+        """
+        root_resource = await self._get_root_resource(resource)
+        return self._get_script(root_resource.get_id(), target_type)
+
+    async def _get_root_resource(self, resource: Resource) -> Resource:
+        if resource.get_id() in self.root_cache:
+            return self.root_cache[resource.get_id()]
+        while len(list(await resource.get_ancestors())) != 0:
+            resource = await resource.get_parent()
+        self.root_cache[resource.get_id()] = resource
+        return resource
+
+    async def _get_variable_from_session(self, resource: Resource):
+        root_resource = await self._get_root_resource(resource)
+        return self.script_sessions[root_resource.get_id()].resource_variable_names[
+            resource.get_id()
+        ]
+
+    async def _var_exists(self, resource: Resource):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource)
+        return resource.get_id() in session.resource_variable_names
+
+    async def _add_action_to_session(self, resource, action, action_type):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        # TODO: actions are duplicated if page is refreshed, is this reasonable?
```

Comment:
This is not reasonable in my opinion. Why does this happen?

I also don't see this happening in practice. Is this still true?

Reply:
I did not see it happening as of yesterday, I just hadn't committed the change to remove it since it was the only change I had at the time and expected that more would need to be made.

Reply:
To answer the first question: it is likely, although I don't recall exactly which actions I was seeing duplicated at the time that I made this comment, that the actions being duplicated were the ones that occur strictly for the frontend to work and don't need to be recorded in the script (e.g., get_data, get_children, etc). Once we removed those, this issue was addressed as well.

Reply:
Cool. If this is no longer an issue, feel free to remove the `TODO` and resolve this thread. Thanks!